File: android/app/src/main/java/com/ventouxlabs/gatepath/diag/VpnProbe.kt
package com.ventouxlabs.gatepath.diag

/**
 * Reads [ProbeContext.vpnInterfaces] / [ProbeContext.isTailscaleFullTunnel]
 * and returns [DiagnosticReport.VpnBlocking] when a VPN is up while the
 * captive portal is unresolved.
 *
 * Any VPN interface is reported — even split-tunnel setups routinely install
 * DNS rules that break captive resolution, so the finding is worth surfacing;
 * [DiagnosticReport.VpnBlocking.isFullTunnel] tells the UI how certain the
 * "pause your VPN" advice is. A Tailscale exit node without a matched
 * interface name (interface enumeration can race teardown) still reports,
 * with a fallback name.
 *
 * No network call — completes synchronously inside the per-probe budget.
 */
class VpnProbe : DiagnosticProbe {
    override val name = "vpn"

    override suspend fun run(ctx: ProbeContext): DiagnosticReport {
        val interfaceName = ctx.vpnInterfaces.firstOrNull()
            ?: if (ctx.isTailscaleFullTunnel) "tailscale" else return DiagnosticReport.Healthy
        return DiagnosticReport.VpnBlocking(
            interfaceName = interfaceName,
            isFullTunnel = ctx.isTailscaleFullTunnel,
        )
    }
}

File: desktop/gatepath/diag/vpn_probe.py
"""Reports a VPN that is likely blocking captive sign-in.

Any VPN interface is reported: even split-tunnel setups routinely install
DNS rules that break captive resolution, so the finding is worth surfacing.
`is_full_tunnel` tells the UI how certain the "pause your VPN" advice is.

Mirror of Android `VpnProbe.kt`, with one structural divergence: Android
falls back to the name "tailscale" when an exit node is active but interface
enumeration matched nothing. Desktop's [ProbeContext] carries full-tunnel
status only per-interface (inside `VpnDetail`), so an empty `vpn_interfaces`
tuple cannot encode that case and the fallback has no data to key off.

Context-only — no network access.
"""

from __future__ import annotations

from gatepath.diag.probe import ProbeContext
from gatepath.diag.report import DiagnosticReport, Healthy, VpnBlocking


class VpnProbe:
    name = "vpn"

    def run(self, ctx: ProbeContext) -> DiagnosticReport:
        if not ctx.vpn_interfaces:
            return Healthy()
        first = ctx.vpn_interfaces[0]
        return VpnBlocking(interface_name=first.name, is_full_tunnel=first.is_full_tunnel)
